relay-redis/src/pool.rs

```diff
@@ -15,6 +17,92 @@ pub type CustomClusterPool = Pool<CustomClusterManager, CustomClusterConnection>
 /// A connection pool for single Redis instance deployments.
 pub type CustomSinglePool = Pool<CustomSingleManager, CustomSingleConnection>;
 
+/// A wrapper for a connection that can be tracked with metadata.
+pub struct TrackedConnection<C> {
+    connection: C,
+    detach: bool,
+}
+
+impl<C> TrackedConnection<C> {
+    fn new(connection: C) -> Self {
+        Self {
+            connection,
+            detach: false,
+        }
+    }
+
+    /// Returns `true` when a [`RedisError`] should lead to the [`TrackedConnection`] be detached
+    /// from the pool.
+    ///
+    /// This is done since some errors can be recoverable, in that case, we want to detach the
+    /// connection to let the pool use or create another one.
+    fn should_be_detached(error: &RedisError) -> bool {
+        !error.is_unrecoverable_error()
+    }
+}
+
+impl<C: redis::aio::ConnectionLike + Send> redis::aio::ConnectionLike for TrackedConnection<C> {
+    fn req_packed_command<'a>(&'a mut self, cmd: &'a Cmd) -> RedisFuture<'a, Value> {
+        async move {
+            let result = self.connection.req_packed_command(cmd).await;
+            if let Err(error) = &result {
+                self.detach = Self::should_be_detached(error);
+            } else {
+                self.detach = false;
+            }
+
+            result
+        }
+        .boxed()
+    }
+
+    fn req_packed_commands<'a>(
+        &'a mut self,
+        cmd: &'a Pipeline,
+        offset: usize,
+        count: usize,
+    ) -> RedisFuture<'a, Vec<Value>> {
+        async move {
+            let result = self
+                .connection
+                .req_packed_commands(cmd, offset, count)
+                .await;
+            if let Err(error) = &result {
+                self.detach = Self::should_be_detached(error);
+            } else {
+                self.detach = false;
```

Comment:
This resetting should not be needed since the recycling should drop the connection but to be safe, we mark the connection as non detachable in case of success.